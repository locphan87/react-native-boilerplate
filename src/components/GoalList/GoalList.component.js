// @flow
import React from 'react'
import { View, FlatList } from 'react-native'
import { compose, prop } from 'ramda'

import GoalItem from '../GoalItem/GoalItem.component'
import type { Goal } from '../../types'

import styles from './GoalList.component.style'

type Props = {
  list: Goal[],
  onPress: Function
}

const keyExtractor = compose(
  String,
  prop('id')
)

const GoalList = ({ list, onPress }: Props) => (
  <View style={styles.container}>
    <FlatList
      data={list}
      keyExtractor={keyExtractor}
      renderItem={({ item }) => <GoalItem {...item} onPress={onPress} />}
    />
  </View>
)
GoalList.displayName = 'GoalList'

export default GoalList
