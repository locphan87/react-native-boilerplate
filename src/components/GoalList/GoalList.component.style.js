import { StyleSheet } from 'react-native'

import { COLORS } from '../../themes'

export default StyleSheet.create({
  container: {
    flex: 1,
    width: '100%',
    marginTop: 0,
    borderTopWidth: 0,
    backgroundColor: COLORS.base
  }
})
