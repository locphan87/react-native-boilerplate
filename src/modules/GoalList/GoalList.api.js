// @flow
import API from '../../services/API.service'
import type { Payload } from '../../types'
import { delay } from '../../utils/common.util'

const getGoalList = async () => {
  await delay(1500)

  return API.get('goals')
}
const addGoal = (payload: Payload) => API.post('goals', payload)
const removeGoal = (id: number) => API.delete(`goals/${id}`)

export default {
  getGoalList,
  addGoal,
  removeGoal
}
