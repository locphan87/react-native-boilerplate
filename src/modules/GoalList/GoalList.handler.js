// @flow
import { Routes } from '../Navigation/Navigation.constant'
import type { ID } from '../../types'

type Props = {
  navigation: Object
}

const handlers = {
  OPEN_DETAILS: (props: Props) => (id: ID) => () => {
    const {
      navigation: { navigate }
    } = props
    navigate(Routes.Goal.GoalDetails, { id })
  }
}

export default handlers
