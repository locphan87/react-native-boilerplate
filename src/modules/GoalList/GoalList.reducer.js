// @flow
import { createActions } from 'redux-actions'
import typeToReducer from 'type-to-reducer'
import Immutable from 'seamless-immutable'

import type { Goal } from '../../types'

import API from './GoalList.api'

type State = {
  list: Goal[],
  loading: boolean
}

const actions = createActions({
  ADD_NEW_GOAL: API.addGoal,
  GET_GOAL_LIST: API.getGoalList,
  REMOVE_GOAL: [API.removeGoal, id => ({ id })]
})

const initialState: State = Immutable({
  list: [],
  loading: false
})

const reducer = typeToReducer(
  {
    [actions.getGoalList]: {
      LOADING: state => ({
        ...state,
        loading: true
      }),
      SUCCESS: (state: State, action): State => ({
        ...state,
        loading: false,
        list: action.payload.data
      }),
      ERROR: state => ({
        ...state,
        loading: false
      })
    },
    [actions.addNewGoal]: {
      SUCCESS: (state: State, { payload }): State => ({
        ...state,
        list: [payload.data, ...state.list]
      })
    },
    [actions.removeGoal]: {
      SUCCESS: (state: State, { meta }): State => {
        const keepItem = item => item.id !== meta.id
        const newGoalList = state.list.filter(keepItem)

        return {
          ...state,
          list: newGoalList
        }
      }
    }
  },
  initialState
)

export { actions, initialState }
export default reducer
