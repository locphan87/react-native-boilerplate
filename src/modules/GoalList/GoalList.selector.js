import { createSelector } from 'reselect'

const goalList = state => state.goal.list
const getGoalById = createSelector([goalList], list => id =>
  list.find(item => item.id === id)
)

export default { goalList, getGoalById }
