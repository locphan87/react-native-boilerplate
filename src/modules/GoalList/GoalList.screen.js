// @flow
import React from 'react'
import { Button } from 'react-native'
import { compose, setStatic, lifecycle, withHandlers } from 'recompose'
import { connect } from 'react-redux'

import { withApp } from '../../hoc'
import I18n from '../../i18n'
import { Routes } from '../Navigation/Navigation.constant'

import GoalListView from './GoalList.view'
import { actions } from './GoalList.reducer'
import handlers from './GoalList.handler'

const mapStateToProps = (state): Object => ({
  loading: state.goal.loading,
  goalList: state.goal.list
})
const mapDispatchToProps = {
  getGoalList: actions.getGoalList
}
const navigationOptions = ({ navigation }) => ({
  headerTitle: I18n.t('goalList.screen.title'),
  tabBarLabel: I18n.t('goalList.screen.title'),
  headerRight: (
    <Button
      title={I18n.t('goalList.screen.newGoal')}
      onPress={() => {
        navigation.navigate(Routes.Goal.GoalCreation)
      }}
    />
  )
})
const callbacks = {
  componentDidMount() {
    this.props.getGoalList()
  }
}

export default compose(
  connect(
    mapStateToProps,
    mapDispatchToProps
  ),
  setStatic('navigationOptions', navigationOptions),
  lifecycle(callbacks),
  withApp({ loading: true }),
  withHandlers(handlers)
)(GoalListView)
