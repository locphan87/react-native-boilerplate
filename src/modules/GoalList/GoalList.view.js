// @flow
import React from 'react'

import GoalList from '../../components/GoalList/GoalList.component'
import { Wrapper } from '../../components/Styled'
import type { Goal } from '../../types'

type Props = {
  OPEN_DETAILS: Function,
  goalList: Goal[]
}

const GoalListView = ({ OPEN_DETAILS, goalList }: Props) => (
  <Wrapper>
    <GoalList list={goalList} onPress={OPEN_DETAILS} />
  </Wrapper>
)
GoalListView.displayName = 'GoalListView'

export default GoalListView
