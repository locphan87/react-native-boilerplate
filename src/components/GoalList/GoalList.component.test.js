import React from 'react'
import { shallow } from 'enzyme'

import { testSnapshots, singleSnapTest } from '../../utils/test.util'

import GoalList from './GoalList.component'

const item = {
  id: 1,
  title: 'title',
  start: 1,
  current: 1,
  end: 2
}
const props = {
  list: [item],
  onPress: id => jest.fn()
}
const wrapper = shallow(<GoalList {...props} />)

describe('GoalList', () => {
  testSnapshots(GoalList, [
    {
      props,
      description: 'basic render'
    }
  ])

  describe('sub render', () => {
    const List = wrapper.find('FlatList')
    const Item = List.props().renderItem({ item })

    singleSnapTest(Item, 'render goal item correctly')
  })
})
