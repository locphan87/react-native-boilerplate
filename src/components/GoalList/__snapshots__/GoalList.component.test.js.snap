// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GoalList snapshots basic render 1`] = `
<List
  containerStyle={
    Object {
      "backgroundColor": "rgba(31, 32, 39, 1)",
      "borderTopWidth": 0,
      "flex": 1,
      "marginTop": 0,
      "width": "100%",
    }
  }
>
  <FlatList
    data={
      Array [
        Object {
          "current": 1,
          "end": 2,
          "id": 1,
          "start": 1,
          "title": "title",
        },
      ]
    }
    disableVirtualization={false}
    horizontal={false}
    initialNumToRender={10}
    keyExtractor={[Function]}
    maxToRenderPerBatch={10}
    numColumns={1}
    onEndReachedThreshold={2}
    renderItem={[Function]}
    scrollEventThrottle={50}
    updateCellsBatchingPeriod={50}
    windowSize={21}
  />
</List>
`;

exports[`GoalList sub render render goal item correctly 1`] = `
<GoalItem
  current={1}
  end={2}
  id={1}
  onPress={[Function]}
  start={1}
  title="title"
/>
`;
